File: boardview.rb
class BoardView < Qt::Widget
  signals 'message(QString)'
  signals 'computer_move()'
  PANEL = 'panel.svg'
  HIT = 'hit.svg'
  MISS = 'miss.svg'
  SHIPS = ['carrier.svg', 'battleship.svg', 'destroyer.svg',
           'submarine.svg', 'patrol boat.svg']
  def initialize(player, visible, parent = nil)
    super parent
    @player = player
    @visible = visible
    @panel = Qt::SvgRenderer.new PANEL
    @hit = Qt::SvgRenderer.new HIT
    @miss = Qt::SvgRenderer.new MISS
    @ship = SHIPS.collect { |filename| Qt::SvgRenderer.new filename }
    @moving, @x0, @y0, @dx, @dy = nil, 0, 0, 0, 0
  end
  def board=(value)
    @player = value
    update
  end
  def mousePressEvent(e)
    w, h = width / Player::N, height / Player::N
    bx, by = e.x / w, e.y / h
    if @visible and @player.placing?
      ship = @player.ship_at(bx, by)
      if ship
        if e.button == Qt::LeftButton
          @moving = ship
          @x0, @y0 = e.x, e.y
        else
          x, y, vertical = *@player.ship(ship)
          dx, dy = bx - x, by - y
          if @player.place ship, x + dx - dy, y + dy - dx, !vertical
            emit message("#{Player::TITLE[ship]} rotated")
          else
            emit message('Invalid placement')
          end
          update
        end
      end
    else
      if not @visible and @player.board(bx, by) == :unknown and not @player.game_over?
        @player.target bx, by
        update
        emit computer_move
      end
    end
  end
  def mouseMoveEvent(e)
    if @moving
      @dx, @dy = e.x - @x0, e.y - @y0
      update
    end
  end
  def mouseReleaseEvent(e)
    if @moving
      w, h = width / Player::N, height / Player::N
      x, y, vertical = *@player.ship(@moving)
      x += (@dx + w / 2) / w
      y += (@dy + h / 2) / h
      if @player.place @moving, x, y, vertical
        emit message("#{Player::TITLE[@moving]} placed")
      else
        emit message('Invalid placement')
      end
      @moving, @x0, @y0, @dx, @dy = nil, 0, 0, 0, 0
      update
    end
  end
  def paintEvent(e)
    w, h = width / Player::N, height / Player::N
    painter = Qt::Painter.new self
    for j in 0 ... Player::N
      for i in 0 ... Player::N
        @panel.render painter, Qt::RectF.new(i * w, j * h, w, h)
      end
    end
    if @visible
      @ship.each_with_index do |ship,i|
        x, y, vertical = *@player.ship(i)
        unless i == @moving
          dx, dy = 0, 0
        else
          dx, dy = @dx, @dy
        end
        unless vertical
          painter.setTransform Qt::Transform.new
          ship.render painter,
                      Qt::RectF.new(x * w + dx, y * h + dy, w * Player::LENGTH[i], h)
        else
          painter.setTransform Qt::Transform.new 0, 1, 1, 0, 0, 0
          ship.render painter,
                      Qt::RectF.new(y * h + dy, x * w + dx, h * Player::LENGTH[i], h)
        end
      end
    end
    for j in 0 ... Player::N
      for i in 0 ... Player::N
        state = @player.board i, j
        if state == :hit
          @hit.render painter, Qt::RectF.new(i * w, j * h, w, h)
        elsif state == :miss
          @miss.render painter, Qt::RectF.new(i * w, j * h, w, h)
        end
      end
    end
  end
end

File: boardview.cc
#include <QtGui/QPainter>
#include <QtGui/QMouseEvent>
#include "boardview.hh"

using namespace std;

#define PANELFILE ":/images/panel.svg"
#define HITFILE ":/images/hit.svg"
#define MISSFILE ":/images/miss.svg"
static const char *SHIPFILE[5] = {":/images/carrier.svg", ":/images/battleship.svg",
                                  ":/images/destroyer.svg", ":/images/submarine.svg",
                                  ":/images/patrol boat.svg"};

BoardView::BoardView(Player *player, bool visible, QWidget *parent):
  QWidget(parent), m_player(player), m_visible(visible)
{
  m_panel = new QSvgRenderer(QString(PANELFILE), this);
  m_hit = new QSvgRenderer(QString(HITFILE), this);
  m_miss = new QSvgRenderer(QString(MISSFILE), this);
  for (int i=0; i<5; i++)
    m_ship.push_back(new QSvgRenderer(QString(SHIPFILE[i])));
  m_moving = -1;
  m_x0 = 0;
  m_y0 = 0;
  m_dx = 0;
  m_dy = 0;
}

void BoardView::setBoard(Player *player)
{
  m_player = player;
  update();
}

void BoardView::paintEvent(QPaintEvent *e)
{
  int w = width() / N;
  int h = height() / N;
  QPainter *painter = new QPainter(this);
  for (int j=0; j<N; j++)
    for (int i=0; i<N; i++)
      m_panel->render(painter, QRectF(i * w, j * h, w, h));
  if (m_visible) {
    for (int i=0; i<COUNT; i++) {
      QSvgRenderer *ship = m_ship[i];
      int x = m_player->ship(i).x;
      int y = m_player->ship(i).y;
      bool vertical = m_player->ship(i).vertical;
      int dx, dy;
      if (i != m_moving) {
        dx = 0;
        dy = 0;
      } else {
        dx = m_dx;
        dy = m_dy;
      };
      if (!vertical) {
        QTransform transform;
        painter->setTransform(transform);
        ship->render(painter, QRectF(x * w + dx, y * h + dy, w * LENGTH[i], h));
      } else {
        QTransform transform(0, 1, 1, 0, 0, 0);
        painter->setTransform(transform);
        ship->render(painter, QRectF(y * h + dy, x * w + dx, h * LENGTH[i], w));
      };
    };
  };
  for (int j=0; j<N; j++)
    for (int i=0; i<N; i++) {
      int state = m_player->board(i, j);
      if (state == HIT)
        m_hit->render(painter, QRectF(i * w, j * h, w, h));
      else if (state == MISS)
        m_miss->render(painter, QRectF(i * w, j * h, w, h));
    };
  delete painter;
}

void BoardView::mousePressEvent(QMouseEvent *e)
{
  int w = width() / N;
  int h = height() / N;
  int bx = e->x() / w;
  int by = e->y() / h;
  if (m_visible && m_player->placing()) {
    int ship = m_player->shipAt(bx, by);
    if (ship != -1) {
      if (e->button() == Qt::LeftButton) {
        m_moving = ship;
        m_x0 = e->x();
        m_y0 = e->y();
      } else {
        int x = m_player->ship(ship).x;
        int y = m_player->ship(ship).y;
        bool vertical = m_player->ship(ship).vertical;
        int dx = bx - x;
        int dy = by - y;
        if (m_player->place(ship, x + dx - dy, y + dy - dx, !vertical)) {
          emit message("Ship rotated");
        } else {
          emit message("Invalid placement");
        };
        update();
      };
    };
  } else {
    if (!m_visible && m_player->board(bx, by) == UNKNOWN && !m_player->gameOver()) {
      m_player->target(bx, by);
      update();
      emit computerMove();
    };
  };
}

void BoardView::mouseMoveEvent(QMouseEvent *e)
{
  if (m_moving != -1) {
    m_dx = e->x() - m_x0;
    m_dy = e->y() - m_y0;
    update();
  };
}

void BoardView::mouseReleaseEvent(QMouseEvent *e)
{
  if (m_moving != -1) {
    int w = width() / N;
    int h = height() / N;
    int x = m_player->ship(m_moving).x;
    int y = m_player->ship(m_moving).y;
    bool vertical = m_player->ship(m_moving).vertical;
    x += (m_dx + w / 2) / w;
    y += (m_dy + h / 2) / h;
    if (m_player->place(m_moving, x, y, vertical)) {
      emit message("Ship placed");
    } else {
      emit message("Invalid placement");
    };
    m_moving = -1;
    m_x0 = 0;
    m_y0 = 0;
    m_dx = 0;
    m_dy = 0;
    update();
  };
}

